### ISSUE TITLE
vcpkg command port-name should be consistent across commands

### ISSUE THREAD

**Issue by BillyONeal** [microsoft/vcpkg#18967]
The pull request series:

* https://github.com/microsoft/vcpkg-tool/pull/56
* https://github.com/microsoft/vcpkg-tool/pull/62
* https://github.com/microsoft/vcpkg-tool/pull/55
* https://github.com/microsoft/vcpkg-tool/pull/93

opened a lot of internal discussion among the vcpkg maintainers. Some maintainers want to make sure registry-ish commands always respect registry behavior, some value not breaking backcompat on any commands above all else, some want to type the fewest characters, etc.

To that end we are suggesting the following behavior:

> `./vcpkg <command> <port-name>` shall have the following behavior:
>  * If there exists a `versions/baseline.json` directory in a directory above this one, then:
>    * let `<registry-root>` be the directory which contains versions.
>    * Any modifications to a port directory or reads of a port directory will use `<registry-root>/ports/<port-name>`.
>    * Modifications to the versions tree will use `<registry-root>/versions`
> * Otherwise, if there exists a `vcpkg.json` file in a directory above this one, then:
>    * Error out until we implement this mode. We probably want to do something that considers any/all registries configured for that manifest.
>  * Otherwise, if the builtin registry is is a filesystem registry, and `vcpkg.exe` is in the root directory of that registry, and that registry contains the file `.vcpkg-root`
>    * Any modifications to a port directory or reads of a port directory will use `<vcpkg-root>/ports/<port-name>`.
>    * Modifications to the versions tree will use `<vcpkg-root>/versions`
>    * Note: this behavior is intended to be a backwards compatible behavior for the existing vcpkg deployment paradigm and is intended to not engage when we make vcpkg acquirable independently of the builtin registry
> * Otherwise, report an error that the port name could not be resolved.
>
> On a code basis, the implementation of the above rules needs to be sufficiently separate from each command that it does not encourage commands to differ in how they respond to port names. An input to the resolution process needs to be whether it is a command intending to 'edit' the ports in question, and to emit an error if a port is selected in an uneditable source, such as a git registry (which vcpkg clones into a temporary location meaning such edits will be lost).
>
> We will not add new commands where it is possible to be ambiguous between a file path and a port, so the outstanding PRs trying to make format-manifest accept port names will be closed in favor of a request to implement a new command consistent with the above behavior, tentatively bikeshedded as `format-port`.

**Comment by dg0yt** [comment]
I  miss a mention of `vcpkg-configuration.json`.
It's not clear to me if the design covers the case where `vcpkg-configuration.json` in an otherwise pristine vpkg root denominates a custom git registry as default registry, and the user wants to do maintain that custom git repository. IMO command should apply to the *default* git registry but this isn't the case at the moment.

FTR, my current workaround for `x-add-version` is:
~~~cmake
${vcpkg} x-add-version "${PORT}"
        --skip-formatting-check
        ${overwrite_option}
        "--x-builtin-ports-root=${CUSTOM_REGISTRY}/ports"
        "--x-builtin-registry-versions-dir=${CUSTOM_REGISTRY}/versions"
~~~

**Comment by autoantwort** [comment]
Can we specify what `<command>` is. I assume it is only `edit`, `x-add-version` and `format-manifest`? So "registry management" commands. Why should these look at a existing `vcpkg.json` file?

> FTR, my current workaround for x-add-version is

With https://github.com/microsoft/vcpkg-tool/pull/93 it would be possible to simply be in the dir of the custom registry or setting it with `--x-registry-root`.

**Comment by BillyONeal** [comment]
> I miss a mention of `vcpkg-configuration.json`.

That's covered by the "Otherwise, if there exists a vcpkg.json file in a directory above this one," line, since to be meaningful a vcpkg-configuration.json needs an associated vcpkg.json. (At least, that is the intent)

We're choosing 'error' as the resolution right now because we could not arrive at a set of behavior that actually made everyone happy, and want to find out from actual customers what they expect in such cases.

> Can we specify what `<command>` is. I assume it is only `edit`, `x-add-version` and `format-manifest`? So "registry management" commands. Why should these look at a existing `vcpkg.json` file?

All commands that speak about port names. `install`, `find`, `x-set-installed`, `build` and friends should also follow the same rules, for example. We value the same name meaning the same thing across commands over doing specific things that might make sense for individual commands, in the same way that git lets you put a tree-ish anywhere a tree-ish is accepted even if the action may be impossible for a given command.

Contrast this with our current 'wild west' of behavior where the means by which commands decide ports are indicated is basically at the whim of the person who implemented that individual command. That code review series where 3 different maintainers expected the port selection process to follow 3 different sets of rules really highlighted that we need a 'how that works' solution.

**Comment by autoantwort** [comment]
> All commands that speak about port names. `install`, `find`, `x-set-installed`, `build` and friends should also follow the same rules, for example.

I disagree. Imo we should really split the behavior of finding ports in two groups, "registry management" and "user" commands. For `x-add-version` it makes no sense to search outside of the current registry that you want to edit. Also I don't want that `format-manifest` searches for a `vcpkg-configuration.json` and then tries to format some files somewhere. When my cwd is inside a custom registry and i use "registry management", I want that these commands only change the registry I am currently in and nothing else.

Additionally I don't think that "user" commands should work inside a custom registry. If a user wants to use custom registry, the user should create a vcpkg.json and `vcpkg-configuration.json` file. Or at least it should then not search for a `vcpkg.json` file and a `vcpkg-configuration.json` when it detects a custom registry. I would find it very strange when I am in a custom registry, want to install some port that does not exists, it then finds some vcpkg.json file up the file tree which has a custom registry pointing to some github repo and it then installs the package. 

Additionally we should maybe make the "registry management" commands for the builtin registry only available inside the builtin registry folder like for custom registries to create a unified experience.

**Comment by BillyONeal** [comment]
>For x-add-version it makes no sense to search outside of the current registry that you want to edit.

And that is why the first option to try is the 'are we in a registry, if so assume the operation targets that registry' check.

>Also I don't want that format-manifest searches for a vcpkg-configuration.json and then tries to format some files somewhere.

It may be that some commands return an error depending on the path by which they find the thing they are talking about; that doesn't mean that the same name on the same command line to the same tool should denote a different entity.

> I would find it very strange when I am in a custom registry, want to install some port that does not exists, it then finds some vcpkg.json file up the file tree

It wouldn't do that under the proposed behavior, because it would find that you were in a registry first, and take that path.

>Additionally we should maybe make the "registry management" commands for the builtin registry only available inside the builtin registry folder like for custom registries to create a unified experience.

That would break backwards compatibility for existing customers. It's currently completely reasonable for someone to put a vcpkg install on their PATH, and assume any mention of the name 'vcpkg' refers to the one they put on their path.

**Comment by autoantwort** [comment]
I think I misunderstood your search description. I thought when a port is not found in the current "level", it continuous searching in the next one. But if I now understand correctly we stop searching on the other "levels" when we found one of the port sources (for example a custom registry).

**Comment by autoantwort** [comment]
https://github.com/microsoft/vcpkg-tool/pull/93 then already partly implements this behavior for `x-add-version`, `civerifyportversions`, `create` and `format-manifest` and it would be very easy (one liner) to do it also for `edit`, `portsdiff`, `autocomplete`. The only think it currently doesn't respect are vcpkg.json files (so creating an error except for `edit` and `autocomplete`).  
Should `install` work inside of a custom registry or should that be an error? Because it installs into `[vcpkg-root]/installed` by default, but ok, you can set `VCPKG_INSTALL_DIR`.

**Comment by BillyONeal** [comment]
> I think I misunderstood your search description. I thought when a port is not found in the current "level", it continuous searching in the next one. But if I now understand correctly we stop searching on the other "levels" when we found one of the port sources (for example a custom registry).

Right, if we enter any of the bullets and find no match we don't continue elsewhere under this proposal.

**Comment by ras0219-msft** [comment]
First suggestion:

>Otherwise, if the builtin registry is is a filesystem registry...

I believe this is attempting to match the non-baseline case for the builtin registry. If so, I do not believe this should use the language "filesystem registry" because that's a well-defined term for users and is not how the non-baseline mode is implemented. Instead, I would say something like

> Otherwise, if the vcpkg tool binary is in the root directory of a vcpkg instance,

Second suggestion:

If the user is not inside the found registry (e.g. it isn't Clause 1), then we should treat any determined registry to be "read-only" and refuse to perform modifying operations. This is a breaking change to `x-add-version`, but avoids spooky action at a distance. It would also be a breaking change for `edit` however, so we may want to keep a carve-out for that specific command for backwards compatibility (perhaps adding a warning to the command line that the user is editing a file that isn't related to their CWD).

**Comment by autoantwort** [comment]
> If the user is not inside the found registry (e.g. it isn't Clause 1), then we should treat any determined registry to be "read-only" and refuse to perform modifying operations.

Then we could remove all other clauses: `edit`, `x-add-version` or `format-manifest` makes no sense in a read only registry. 

In general. Should these commands support filesystem registries? Currently they only support git registries and without a default baseline it is afaik impossible to implement then for filesystem registries. Imho these commands should not support filesystem registries.

**Comment by theblackunknown** [comment]
Hi I am bumping this thread.
This proposal looks great, what is the status on this ?
As a maintainer, I am interested into microsoft/vcpkg-tool#93 but this PR behaviour clarification is required before it can proceed further.

**Comment by BillyONeal** [comment]
@theblackunknown Other than that the above resolution is still what we want nothing much has changed here; we just haven't gotten to it yet

**Comment by theblackunknown** [comment]
Thanks for the update @BillyONeal , is there anything I can do to help ?
For example is there something to implement for this resolution which I can try to implement ?

**Comment by BillyONeal** [comment]
> Thanks for the update @BillyONeal , is there anything I can do to help ? For example is there something to implement for this resolution which I can try to implement ?

I don't think anyone is currently working on that so if you want to try to fix x-add-version and/or format-manifest to be consistent with the above plan that would be a start.

**Comment by theblackunknown** [comment]
As suggested, I have opened a draft PR to implement the wanted behavior and get more feedbacks as a first time contributor to this code base.
Can someone get a look at it and let me know whether this is going into the right direction or not ?
cf. microsoft/vcpkg-tool#517
